File: sol/3494/3494.py
class Solution:
    def minTime(self, skill: List[int], mana: List[int]) -> int:
        n = len(skill)
        freeTime = [0] * n
        baseDelay = [0] * (n + 1)
        for i, sk in enumerate(skill):
            baseDelay[i + 1] = baseDelay[i] + sk
        for m in mana:
            minStart = max(0, max(a - m * b for a, b in zip(freeTime, baseDelay)))
            for i in range(n):
                freeTime[i] = minStart + m * baseDelay[i + 1]
        return freeTime[-1]

File: sol/3494/3494.cpp
#include <ranges>

struct Solution {
    using ll = long long;
    long long minTime(vector<int>& skill, vector<int>& mana) {
        int n = skill.size();
        vector<ll> freeTime(n, 0), baseDelay(n + 1, 0);
        for (int i = 0; i < n; ++i){
            baseDelay[i + 1] = baseDelay[i] + skill[i];
        }
        for (int m : mana){
            ll minStart = 0;
            for (auto [a, b] : ranges::views::zip(freeTime, baseDelay)){
                minStart = max(minStart, a - m * b);
            }
            for (int i = 0; i < n; ++i){
                freeTime[i] = minStart + m * baseDelay[i + 1];
            }
        }
        return freeTime.back();
    }
};
